RawSpeed/DngDecoderSlices.cpp

```diff
@@ -143,6 +143,204 @@ my_error_throw (j_common_ptr cinfo)
+  // binary32 equation: -1 ^ sign * 2 ^ (exponent - 127) * 1.fraction
+  // => exponent32 - 127 = exponent24 - 64, exponent32 = exponent16 + 127 - 63
+  uint32_t fp32_exponent = fp24_exponent + 127 - 63;
+  uint32_t fp32_fraction = fp24_fraction << (23 - 16); // 23 is binary 32 fraction size
+
+  if (fp24_exponent == 127) {
+    // Infinity or NaN
+    fp32_exponent = 255;
+  } else if (fp24_exponent == 0 && fp24_fraction == 0) {
+    // +-Zero
+    fp32_exponent = 0;
+    fp32_fraction = 0;
+  } else if (fp24_exponent == 0 && fp24_fraction != 0) {
+    // Subnormal numbers
+    // binary32 equation: -1 ^ sign * 2 ^ (exponent - 127) * 1.fraction
+    // binary24 equation: -1 ^ sign * 2 ^ -62 * 0.fraction, we can represent it
+    // as a normalized value in binary32, we have to shift fraction until we get 1.new_fraction
+    // and decrement exponent for each shift
+    fp32_exponent = -62 + 127;
+    while (!(fp32_fraction & (1 << 23))) {
+      fp32_exponent -= 1;
+      fp32_fraction <<= 1;
+    }
+    fp32_fraction &= ((1 << 23) - 1);
+  }
+  return (sign << 31) | (fp32_exponent << 23) | fp32_fraction;
+}
+
+static void expandFP16(unsigned char *dst, int width)
+{
+  uint16_t *dst16 = (uint16_t *)dst;
+  uint32_t *dst32 = (uint32_t *)dst;
+
+  for (off_t x = width - 1; x >= 0; x--)
+    dst32[x] = fp16ToFloat(dst16[x]);
+}
+
+static void expandFP24(unsigned char *dst, int width)
+{
+  uint32_t *dst32 = (uint32_t *)dst;
+  dst += (width - 1) * 3;
+  for (off_t x = width -1; x >= 0; x--) {
+    dst32[x] = fp24ToFloat((dst[0] << 16) | (dst[1] << 8) | dst[2]);
+    dst -= 3;
+  }
+}
+
+void DngDecoderSlices::decodeDeflate(const DngSliceElement &e, unsigned char *uBuffer)
+{
+  if (!mFile->isValid(e.byteOffset, e.byteCount))
+    ThrowRDE("DngDecoderSlices::decodeDeflate: tile offset plus size is longer than file. Truncated file.");
+
+  unsigned long dstLen = e.width * e.height * 4;
+  const unsigned char *cBuffer = mFile->getData(e.byteOffset, e.byteCount);
+  int predFactor = 0;
+
+  switch (mPredictor) {
+  case 3:
+    predFactor = 1;
+    break;
+  case 34894:
+    predFactor = 2;
+    break;
+  case 34895:
+    predFactor = 4;
+    break;
+  default:
+    predFactor = 0;
+    break;
+  }
+
+  int err = uncompress(uBuffer, &dstLen, cBuffer, e.byteCount);
+  if (err != Z_OK)
+    ThrowRDE("DngDecoderSlices::decodeDeflate: failed to uncompress tile: %d", err);
+  int bytesps = mBps / 8;
+  size_t thisTileLength = e.offY + e.height > (uint32)mRaw->dim.y ? mRaw->dim.y - e.offY : e.height;
+  size_t thisTileWidth = e.offX + e.width > (uint32)mRaw->dim.x ? mRaw->dim.x - e.offX : e.width;
+  for (size_t row = 0; row < thisTileLength; ++row) {
+    unsigned char * src = uBuffer + row * e.width * bytesps;
+    unsigned char *dst = (unsigned char *)mRaw->getData() + ((e.offY + row) * mRaw->pitch + e.offX * sizeof(float) * mRaw->getCpp());
+    if (predFactor)
+      decodeFPDeltaRow(src, dst, thisTileWidth, e.width, bytesps, predFactor);
```

Comment:
As @axxel points, regardless of whether it actually helps or not, it would be best if these `decodeFPDeltaRow()`, `expandFP16()`, `expandFP24()`, `fp16ToFloat()` and `fp24ToFloat()` and  be marked with `inline` keyword.